File: cocos/3d/models/skinning-model.ts
/*
 Copyright (c) 2017-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { Material } from '../../asset/assets/material';
import { RenderingSubMesh } from '../../asset/assets/rendering-sub-mesh';
import { Mesh } from '../assets/mesh';
import { Skeleton } from '../assets/skeleton';
import { geometry, Mat4, Vec3, warnID } from '../../core';
import { BufferUsageBit, MemoryUsageBit, DescriptorSet, Buffer, BufferInfo, Attribute, FormatFeatureBit, Format } from '../../gfx';
import { UBOSkinning, UNIFORM_REALTIME_JOINT_TEXTURE_BINDING } from '../../rendering/define';
import { Node } from '../../scene-graph/node';
import { ModelType } from '../../render-scene/scene/model';
import { uploadJointData } from '../skeletal-animation/skeletal-animation-utils';
import { MorphModel } from './morph-model';
import { deleteTransform, getTransform, getWorldMatrix, IJointTransform } from '../../animation/skeletal-animation-utils';
import { IMacroPatch, BatchingSchemes, Pass } from '../../render-scene';
import { director } from '../../game';
import { PixelFormat } from '../../asset/assets/asset-enum';
import { Texture2D, ImageAsset } from '../../asset/assets';
import { SubModel } from '../../render-scene/scene';

const uniformPatches: IMacroPatch[] = [
    { name: 'CC_USE_SKINNING', value: true },
    { name: 'CC_USE_REAL_TIME_JOINT_TEXTURE', value: false },
];
const texturePatches: IMacroPatch[] = [
    { name: 'CC_USE_SKINNING', value: true },
    { name: 'CC_USE_REAL_TIME_JOINT_TEXTURE', value: true },
];

function getRelevantBuffers (outIndices: number[], outBuffers: number[], jointMaps: number[][], targetJoint: number): void {
    for (let i = 0; i < jointMaps.length; i++) {
        const idxMap = jointMaps[i];
        let index = -1;
        for (let j = 0; j < idxMap.length; j++) {
            if (idxMap[j] === targetJoint) { index = j; break; }
        }
        if (index >= 0) {
            outBuffers.push(i);
            outIndices.push(index);
        }
    }
}

interface IJointInfo {
    bound: geometry.AABB;
    target: Node;
    bindpose: Mat4;
    transform: IJointTransform;
    buffers: number[];
    indices: number[];
}

const v3_min = new Vec3();
const v3_max = new Vec3();
const v3_1 = new Vec3();
const v3_2 = new Vec3();
const m4_1 = new Mat4();
const ab_1 = new geometry.AABB();

class RealTimeJointTexture {
    public static readonly WIDTH = 256;
    public static readonly HEIGHT = 3;
    public _format = PixelFormat.RGBA32F; // default use float texture
    public _textures: Texture2D[] = [];
    public _buffers: Float32Array[] = [];
}

/**
 * @en
 * The skinning model that is using real-time pose calculation.
 * @zh
 * 实时计算动画的蒙皮模型。
 */
export class SkinningModel extends MorphModel {
    private _buffers: Buffer[] = [];
    private _dataArray: Float32Array[] = [];
    private _joints: IJointInfo[] = [];
    private _bufferIndices: number[] | null = null;
    private _realTimeJointTexture = new RealTimeJointTexture();
    private _realTimeTextureMode = false;
    constructor () {
        super();
        this.type = ModelType.SKINNING;
    }

    public destroy (): void {
        this.bindSkeleton();
        if (this._buffers.length) {
            for (let i = 0; i < this._buffers.length; i++) {
                this._buffers[i].destroy();
            }
            this._buffers.length = 0;
        }
        this._dataArray.length = 0;
        this._realTimeJointTexture._textures.forEach((tex) => {
            tex.destroy();
        });
        this._realTimeJointTexture._textures.length = 0;
        this._realTimeJointTexture._buffers.length = 0;
        super.destroy();
    }

    /**
     * @en Abstract function for [[BakedSkinningModel]], empty implementation.
     * @zh 由 [[BakedSkinningModel]] 继承的空函数。
     */
    public uploadAnimation () : void {}

    /**
     * @en Bind the skeleton with its skinning root node and the mesh data.
     * @zh 在模型中绑定一个骨骼，需要提供骨骼的蒙皮根节点和蒙皮网格数据。
     * @param skeleton @en The skeleton to be bound @zh 要绑定的骨骼
     * @param skinningRoot @en The skinning root of the skeleton @zh 骨骼的蒙皮根节点
     * @param mesh @en The mesh @zh 蒙皮网格
     * @returns void
     */
    public bindSkeleton (skeleton: Skeleton | null = null, skinningRoot: Node | null = null, mesh: Mesh | null = null): void {
        for (let i = 0; i < this._joints.length; i++) {
            deleteTransform(this._joints[i].target);
        }
        this._bufferIndices = null; this._joints.length = 0;
        if (!skeleton || !skinningRoot || !mesh) { return; }
        this._realTimeTextureMode = false;
        if (UBOSkinning.JOINT_UNIFORM_CAPACITY < skeleton.joints.length) { this._realTimeTextureMode = true; }
        this.transform = skinningRoot;
        const boneSpaceBounds = mesh.getBoneSpaceBounds(skeleton);
        const jointMaps = mesh.struct.jointMaps;
        this._ensureEnoughBuffers(jointMaps && jointMaps.length || 1);
        this._bufferIndices = mesh.jointBufferIndices;
        this._initRealTimeJointTexture();
        for (let index = 0; index < skeleton.joints.length; index++) {
            const bound = boneSpaceBounds[index];
            const target = skinningRoot.getChildByPath(skeleton.joints[index]);
            if (!bound || !target) { continue; }
            const transform = getTransform(target, skinningRoot)!;
            const bindpose = skeleton.bindposes[index];
            const indices: number[] = [];
            const buffers: number[] = [];
            if (!jointMaps) { indices.push(index); buffers.push(0); } else { getRelevantBuffers(indices, buffers, jointMaps, index); }
            this._joints.push({ indices, buffers, bound, target, bindpose, transform });
        }
    }

    /**
     * @en Update world transform and bounding boxes for the model
     * @zh 更新模型的世界矩阵和包围盒
     * @param stamp @en The update time stamp @zh 更新的时间戳
     */
    public updateTransform (stamp: number): void {
        const root = this.transform;
        if (root.hasChangedFlags || root.isTransformDirty()) {
            root.updateWorldTransform();
            this._localDataUpdated = true;
        }
        // update bounds
        Vec3.set(v3_min,  Infinity,  Infinity,  Infinity);
        Vec3.set(v3_max, -Infinity, -Infinity, -Infinity);
        for (let i = 0; i < this._joints.length; i++) {
            const { bound, transform } = this._joints[i];
            const worldMatrix = getWorldMatrix(transform, stamp);
            geometry.AABB.transform(ab_1, bound, worldMatrix);
            ab_1.getBoundary(v3_1, v3_2);
            Vec3.min(v3_min, v3_min, v3_1);
            Vec3.max(v3_max, v3_max, v3_2);
        }

        const worldBounds = this._worldBounds;
        if (this._modelBounds && worldBounds) {
            geometry.AABB.fromPoints(this._modelBounds, v3_min, v3_max);
            this._modelBounds.transform(root._mat, root._pos, root._rot, root._scale, this._worldBounds!);
        }
    }

    /**
     * @en Update uniform buffer objects for rendering.
     * @zh 更新用于渲染的 UBO
     * @param stamp @en The update time stamp @zh 更新的时间戳
     * @returns @en successful or not @zh 更新是否成功
     */
    public updateUBOs (stamp: number): boolean {
        super.updateUBOs(stamp);
        for (let i = 0; i < this._joints.length; i++) {
            const { indices, buffers, transform, bindpose } = this._joints[i];
            Mat4.multiply(m4_1, transform.world, bindpose);
            for (let b = 0; b < buffers.length; b++) {
                uploadJointData(this._dataArray[buffers[b]], indices[b] * 12, m4_1, i === 0);
            }
        }
        if (this._realTimeTextureMode) {
            this._updateRealTimeJointTextureBuffer();
        } else {
            for (let b = 0; b < this._buffers.length; b++) {
                this._buffers[b].update(this._dataArray[b]);
            }
        }
        return true;
    }

    /**
     * @en Initialize sub model with the sub mesh data and the material
     * @zh 用子网格数据和材质初始化一个子模型
     * @param idx @en The index of the sub model to be initialized @zh 需要初始化的子模型序号
     * @param subMeshData @en The sub mesh data @zh 子网格数据
     * @param mat @en The material @zh 子模型材质
     */
    public initSubModel (idx: number, subMeshData: RenderingSubMesh, mat: Material): void {
        const original = subMeshData.vertexBuffers;
        const iaInfo = subMeshData.iaInfo;
        iaInfo.vertexBuffers = subMeshData.jointMappedBuffers;
        super.initSubModel(idx, subMeshData, mat);
        iaInfo.vertexBuffers = original;
    }

    // override
    public getMacroPatches (subModelIndex: number): IMacroPatch[] | null {
        const superMacroPatches = super.getMacroPatches(subModelIndex);
        let myPatches = uniformPatches;
        if (this._realTimeTextureMode) {
            myPatches = texturePatches;
        }
        if (superMacroPatches) {
            return myPatches.concat(superMacroPatches);
        }
        return myPatches;
    }

    /**
     * @deprecated since v3.5.0, this is an engine private interface that will be removed in the future.
     */
    public _updateLocalDescriptors (submodelIdx: number, descriptorSet: DescriptorSet): void {
        super._updateLocalDescriptors(submodelIdx, descriptorSet);
        const idx = this._bufferIndices![submodelIdx];
        if (this._realTimeTextureMode) {
            this._bindRealTimeJointTexture(idx, descriptorSet);
        } else {
            const buffer = this._buffers[idx];
            if (buffer) { descriptorSet.bindBuffer(UBOSkinning.BINDING, buffer); }
        }
    }

    protected _updateInstancedAttributes (attributes: Attribute[], subModel: SubModel): void {
        const pass = subModel.passes[0];
        if (pass.batchingScheme !== BatchingSchemes.NONE) {
            // TODO(holycanvas): #9203 better to print the complete path instead of only the current node
            warnID(3936, this.node.getPathInHierarchy());
        }
        super._updateInstancedAttributes(attributes, subModel);
    }

    private _ensureEnoughBuffers (count: number): void {
        if (this._buffers.length) {
            for (let i = 0; i < this._buffers.length; i++) {
                this._buffers[i].destroy();
            }
            this._buffers.length = 0;
        }

        if (this._dataArray.length) this._dataArray.length = 0;

        if (!this._realTimeTextureMode) {
            for (let i = 0; i < count; i++) {
                this._buffers[i] = this._device.createBuffer(new BufferInfo(
                    BufferUsageBit.UNIFORM | BufferUsageBit.TRANSFER_DST,
                    MemoryUsageBit.HOST | MemoryUsageBit.DEVICE,
                    UBOSkinning.SIZE,
                    UBOSkinning.SIZE,
                ));
                const maxJoints = UBOSkinning.JOINT_UNIFORM_CAPACITY;
                this._dataArray[i] = new Float32Array(12 * maxJoints);
            }
        } else {
            for (let i = 0; i < count; i++) {
                const maxJoints = RealTimeJointTexture.WIDTH;
                this._dataArray[i] = new Float32Array(12 * maxJoints);
            }
        }
    }

    private _initRealTimeJointTexture (): void {
        if (this._realTimeJointTexture._textures.length) {
            this._realTimeJointTexture._textures.forEach((tex) => {
                tex.destroy();
            });
            this._realTimeJointTexture._textures.length = 0;
        }
        this._realTimeJointTexture._buffers.length = 0;
        if (!this._realTimeTextureMode) return;

        const gfxDevice = director.root!.device;
        let width = RealTimeJointTexture.WIDTH;
        const height = RealTimeJointTexture.HEIGHT;
        const hasFeatureFloatTexture = gfxDevice.getFormatFeatures(Format.RGBA32F) & FormatFeatureBit.SAMPLED_TEXTURE;
        if (hasFeatureFloatTexture === 0) {
            this._realTimeJointTexture._format = PixelFormat.RGBA8888;
            width = 4 * RealTimeJointTexture.WIDTH;
        }

        const textures = this._realTimeJointTexture._textures;
        const buffers = this._realTimeJointTexture._buffers;
        const pixelFormat = this._realTimeJointTexture._format;
        for (let i = 0; i < this._dataArray.length; i++) {
            buffers[i] = new Float32Array(4 * RealTimeJointTexture.HEIGHT * RealTimeJointTexture.WIDTH);
            const arrayBuffer = buffers[i];
            const updateView =  pixelFormat === PixelFormat.RGBA32F ? arrayBuffer : new Uint8Array(arrayBuffer.buffer);
            const image = new ImageAsset({
                width,
                height,
                _data: updateView,
                _compressed: false,
                format: pixelFormat,
            });
            const texture = new Texture2D();
            texture.setFilters(Texture2D.Filter.NEAREST, Texture2D.Filter.NEAREST);
            texture.setMipFilter(Texture2D.Filter.NONE);
            texture.setWrapMode(Texture2D.WrapMode.CLAMP_TO_EDGE, Texture2D.WrapMode.CLAMP_TO_EDGE, Texture2D.WrapMode.CLAMP_TO_EDGE);
            texture.image = image;
            textures[i] = texture;
        }
    }

    private _bindRealTimeJointTexture (idx: number, descriptorSet: DescriptorSet): void {
        if (!this._realTimeTextureMode) return;
        const jointTexture = this._realTimeJointTexture._textures[idx];
        if (jointTexture) {
            const gfxTexture = jointTexture.getGFXTexture();
            const sampler = jointTexture.getGFXSampler();
            descriptorSet.bindTexture(UNIFORM_REALTIME_JOINT_TEXTURE_BINDING, gfxTexture!);
            descriptorSet.bindSampler(UNIFORM_REALTIME_JOINT_TEXTURE_BINDING, sampler);
        }
    }

    private _updateRealTimeJointTextureBuffer (): void {
        if (!this._realTimeTextureMode) return;
        const textures = this._realTimeJointTexture._textures;
        const buffers = this._realTimeJointTexture._buffers;
        for (let idx = 0; idx < textures.length; idx++) {
            const arrayBuffer = buffers[idx];
            const src = this._dataArray[idx];
            const count = src.length / 12; // mat3x4
            let idxSrc = 0;
            let idxDst = 0;
            for (let i = 0; i < count; i++) {
                idxDst = 4 * i;
                arrayBuffer[idxDst++] = src[idxSrc++];
                arrayBuffer[idxDst++] = src[idxSrc++];
                arrayBuffer[idxDst++] = src[idxSrc++];
                arrayBuffer[idxDst++] = src[idxSrc++];
                idxDst = 4 * (i + RealTimeJointTexture.WIDTH);
                arrayBuffer[idxDst++] = src[idxSrc++];
                arrayBuffer[idxDst++] = src[idxSrc++];
                arrayBuffer[idxDst++] = src[idxSrc++];
                arrayBuffer[idxDst++] = src[idxSrc++];
                idxDst = 4 * (i + 2 * RealTimeJointTexture.WIDTH);
                arrayBuffer[idxDst++] = src[idxSrc++];
                arrayBuffer[idxDst++] = src[idxSrc++];
                arrayBuffer[idxDst++] = src[idxSrc++];
                arrayBuffer[idxDst++] = src[idxSrc++];
            }
            const pixelFormat = this._realTimeJointTexture._format;
            const updateView = pixelFormat === PixelFormat.RGBA32F ? arrayBuffer : new Uint8Array(arrayBuffer.buffer);
            textures[idx].uploadData(updateView);
        }
    }
}

File: native/cocos/3d/models/SkinningModel.cpp
/****************************************************************************
 Copyright (c) 2021-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/
#include "3d/models/SkinningModel.h"

#include <utility>

#include "3d/assets/Mesh.h"
#include "3d/assets/Skeleton.h"
#include "core/platform/Debug.h"
#include "core/scene-graph/Node.h"
#include "renderer/gfx-base/GFXBuffer.h"
#include "scene/Pass.h"
#include "scene/RenderScene.h"

const uint32_t REALTIME_JOINT_TEXTURE_WIDTH = 256;
const uint32_t REALTIME_JOINT_TEXTURE_HEIGHT = 3;

namespace {
void getRelevantBuffers(ccstd::vector<index_t> &outIndices, ccstd::vector<int32_t> &outBuffers, const ccstd::vector<ccstd::vector<int32_t>> &jointMaps, int32_t targetJoint) {
    for (int32_t i = 0; i < jointMaps.size(); i++) {
        index_t index = CC_INVALID_INDEX;
        for (int32_t j = 0; j < jointMaps[i].size(); j++) {
            if (jointMaps[i][j] == targetJoint) {
                index = j;
                break;
            }
        }
        if (index >= 0) {
            outBuffers.emplace_back(i);
            outIndices.emplace_back(index);
        }
    }
}

ccstd::vector<cc::scene::IMacroPatch> uniformPatches{{"CC_USE_SKINNING", true}, {"CC_USE_REAL_TIME_JOINT_TEXTURE", false}};
ccstd::vector<cc::scene::IMacroPatch> texturePatches{{"CC_USE_SKINNING", true}, {"CC_USE_REAL_TIME_JOINT_TEXTURE", true}};

} // namespace
namespace cc {

SkinningModel::SkinningModel() {
    _type = Model::Type::SKINNING;
}
SkinningModel::~SkinningModel() {
    releaseData();
}

void SkinningModel::destroy() {
    bindSkeleton(nullptr, nullptr, nullptr);
    releaseData();
    Super::destroy();
}

void SkinningModel::bindSkeleton(Skeleton *skeleton, Node *skinningRoot, Mesh *mesh) {
    for (const JointInfo &joint : _joints) {
        deleteTransform(joint.target);
    }
    _bufferIndices.clear();
    _joints.clear();

    if (!skeleton || !skinningRoot || !mesh) return;
    auto jointCount = static_cast<uint32_t>(skeleton->getJoints().size());
    _realTimeTextureMode = pipeline::SkinningJointCapacity::jointUniformCapacity < jointCount;
    setTransform(skinningRoot);
    auto boneSpaceBounds = mesh->getBoneSpaceBounds(skeleton);
    const auto &jointMaps = mesh->getStruct().jointMaps;
    ensureEnoughBuffers((jointMaps.has_value() && !jointMaps->empty()) ? static_cast<uint32_t>(jointMaps->size()) : 1);
    _bufferIndices = mesh->getJointBufferIndices();
    initRealTimeJointTexture();
    for (index_t index = 0; index < skeleton->getJoints().size(); ++index) {
        geometry::AABB *bound = boneSpaceBounds[index];
        auto *target = skinningRoot->getChildByPath(skeleton->getJoints()[index]);
        if (!bound || !target) continue;

        auto *transform = cc::getTransform(target, skinningRoot);
        const Mat4 &bindPose = skeleton->getBindposes()[index];
        ccstd::vector<index_t> indices;
        ccstd::vector<index_t> buffers;
        if (!jointMaps.has_value()) {
            indices.emplace_back(index);
            buffers.emplace_back(0);
        } else {
            getRelevantBuffers(indices, buffers, jointMaps.value(), index);
        }

        JointInfo jointInfo;
        jointInfo.bound = bound;
        jointInfo.target = target;
        jointInfo.bindpose = bindPose;
        jointInfo.transform = transform;
        jointInfo.buffers = std::move(buffers);
        jointInfo.indices = std::move(indices);
        _joints.emplace_back(std::move(jointInfo));
    }
}

void SkinningModel::updateTransform(uint32_t stamp) {
    auto *root = getTransform();
    if (root->getChangedFlags() || root->isTransformDirty()) {
        root->updateWorldTransform();
        _localDataUpdated = true;
    }
    Vec3 v3Min{INFINITY, INFINITY, INFINITY};
    Vec3 v3Max{-INFINITY, -INFINITY, -INFINITY};
    geometry::AABB ab1;
    Vec3 v31;
    Vec3 v32;
    for (JointInfo &jointInfo : _joints) {
        auto &transform = jointInfo.transform;
        Mat4 worldMatrix = cc::getWorldMatrix(transform, static_cast<int32_t>(stamp));
        jointInfo.bound->transform(worldMatrix, &ab1);
        ab1.getBoundary(&v31, &v32);
        Vec3::min(v3Min, v31, &v3Min);
        Vec3::max(v3Max, v32, &v3Max);
    }
    if (_modelBounds && _modelBounds->isValid() && _worldBounds) {
        geometry::AABB::fromPoints(v3Min, v3Max, _modelBounds);
        _modelBounds->transform(root->getWorldMatrix(), _worldBounds);
        _worldBoundsDirty = true;
    }
}

void SkinningModel::updateUBOs(uint32_t stamp) {
    Super::updateUBOs(stamp);
    uint32_t bIdx = 0;
    Mat4 mat4;
    for (const JointInfo &jointInfo : _joints) {
        Mat4::multiply(jointInfo.transform->world, jointInfo.bindpose, &mat4);
        for (uint32_t buffer : jointInfo.buffers) {
            uploadJointData(jointInfo.indices[bIdx] * 12, mat4, _dataArray[buffer]);
            bIdx++;
        }
        bIdx = 0;
    }
    if (_realTimeTextureMode) {
        updateRealTimeJointTextureBuffer();
    } else {
        bIdx = 0;
        for (gfx::Buffer *buffer : _buffers) {
            buffer->update(_dataArray[bIdx], buffer->getSize());
            bIdx++;
        }
    }
}

void SkinningModel::initSubModel(index_t idx, RenderingSubMesh *subMeshData, Material *mat) {
    const auto &original = subMeshData->getVertexBuffers();
    auto &iaInfo = subMeshData->getIaInfo();
    iaInfo.vertexBuffers = subMeshData->getJointMappedBuffers();
    Super::initSubModel(idx, subMeshData, mat);
    iaInfo.vertexBuffers = original;
}

ccstd::vector<scene::IMacroPatch> SkinningModel::getMacroPatches(index_t subModelIndex) {
    auto patches = Super::getMacroPatches(subModelIndex);
    auto myPatches = uniformPatches;
    if (_realTimeTextureMode) {
        myPatches = texturePatches;
    }
    if (!patches.empty()) {
        patches.reserve(myPatches.size() + patches.size());
        patches.insert(std::begin(patches), std::begin(myPatches), std::end(myPatches));
        return patches;
    }

    return myPatches;
}

void SkinningModel::uploadJointData(uint32_t base, const Mat4 &mat, float *dst) {
    memcpy(reinterpret_cast<void *>(dst + base), mat.m, sizeof(float) * 12);
    dst[base + 3] = mat.m[12];
    dst[base + 7] = mat.m[13];
    dst[base + 11] = mat.m[14];
}

void SkinningModel::updateLocalDescriptors(index_t submodelIdx, gfx::DescriptorSet *descriptorset) {
    Super::updateLocalDescriptors(submodelIdx, descriptorset);
    uint32_t idx = _bufferIndices[submodelIdx];
    if (!_realTimeTextureMode) {
        gfx::Buffer *buffer = _buffers[idx];
        if (buffer) {
            descriptorset->bindBuffer(pipeline::UBOSkinning::BINDING, buffer);
        }
    } else {
        bindRealTimeJointTexture(idx, descriptorset);
    }
}

void SkinningModel::updateInstancedAttributes(const ccstd::vector<gfx::Attribute> &attributes, scene::SubModel *subModel) {
    auto *pass = subModel->getPass(0);
    if (pass->getBatchingScheme() != scene::BatchingSchemes::NONE) {
        // TODO(holycanvas): #9203 better to print the complete path instead of only the current node
        debug::warnID(3936, getNode()->getName());
        CC_LOG_WARNING("pass batchingScheme is none, %s", getNode()->getName().c_str());
    }
    Super::updateInstancedAttributes(attributes, subModel);
}

void SkinningModel::ensureEnoughBuffers(uint32_t count) {
    if (!_buffers.empty()) {
        for (gfx::Buffer *buffer : _buffers) {
            CC_SAFE_DESTROY(buffer);
        }
        _buffers.clear();
    }
    if (!_dataArray.empty()) {
        for (auto *data : _dataArray) {
            CC_SAFE_DELETE_ARRAY(data);
        }
        _dataArray.clear();
    }
    _dataArray.resize(count);
    if (!_realTimeTextureMode) {
        _buffers.resize(count);
        uint32_t length = pipeline::UBOSkinning::count;
        for (uint32_t i = 0; i < count; i++) {
            _buffers[i] = _device->createBuffer({
                gfx::BufferUsageBit::UNIFORM | gfx::BufferUsageBit::TRANSFER_DST,
                gfx::MemoryUsageBit::HOST | gfx::MemoryUsageBit::DEVICE,
                pipeline::UBOSkinning::size,
                pipeline::UBOSkinning::size,
            });
            _dataArray[i] = new float[length];
            memset(_dataArray[i], 0, sizeof(float) * length);
        }
    } else {
        uint32_t length = 4 * REALTIME_JOINT_TEXTURE_WIDTH * REALTIME_JOINT_TEXTURE_HEIGHT;
        for (uint32_t i = 0; i < count; i++) {
            if (_dataArray[i] == nullptr) {
                _dataArray[i] = new float[length];
                memset(_dataArray[i], 0, sizeof(float) * length);
            }
        }
    }
}

void SkinningModel::initRealTimeJointTexture() {
    CC_SAFE_DELETE(_realTimeJointTexture);
    if (!_realTimeTextureMode) return;
    _realTimeJointTexture = ccnew RealTimeJointTexture;
    auto *device = gfx::Device::getInstance();
    uint32_t texWidth = REALTIME_JOINT_TEXTURE_WIDTH;
    uint32_t texHeight = REALTIME_JOINT_TEXTURE_HEIGHT;
    gfx::Format textureFormat = gfx::Format::RGBA32F;

    gfx::FormatFeature formatFeature = device->getFormatFeatures(gfx::Format::RGBA32F);
    if (!(formatFeature & gfx::FormatFeature::SAMPLED_TEXTURE)) {
        textureFormat = gfx::Format::RGBA8;
        texWidth = texWidth * 4;
    }
    uint32_t length = 4 * REALTIME_JOINT_TEXTURE_WIDTH * REALTIME_JOINT_TEXTURE_HEIGHT;
    const size_t count = _dataArray.size();
    for (size_t i = 0; i < count; i++) {
        gfx::TextureInfo textureInfo;
        textureInfo.width = texWidth;
        textureInfo.height = texHeight;
        textureInfo.usage = gfx::TextureUsageBit::STORAGE | gfx::TextureUsageBit::SAMPLED | gfx::TextureUsageBit::TRANSFER_SRC | gfx::TextureUsageBit::TRANSFER_DST;
        textureInfo.format = textureFormat;
        IntrusivePtr<gfx::Texture> texture = device->createTexture(textureInfo);
        _realTimeJointTexture->textures.push_back(texture);
    }
    _realTimeJointTexture->buffer = new float[length];
}

void SkinningModel::bindRealTimeJointTexture(uint32_t idx, gfx::DescriptorSet *descriptorset) {
    if (_realTimeJointTexture->textures.size() < idx + 1) return;
    gfx::Texture *texture = _realTimeJointTexture->textures[idx];
    if (texture) {
        gfx::SamplerInfo info{
            gfx::Filter::POINT,
            gfx::Filter::POINT,
            gfx::Filter::NONE,
            gfx::Address::CLAMP,
            gfx::Address::CLAMP,
            gfx::Address::CLAMP,
        };
        auto *device = gfx::Device::getInstance();
        auto *sampler = device->getSampler(info);
        descriptorset->bindTexture(pipeline::REALTIMEJOINTTEXTURE::BINDING, texture);
        descriptorset->bindSampler(pipeline::REALTIMEJOINTTEXTURE::BINDING, sampler);
    }
}

void SkinningModel::updateRealTimeJointTextureBuffer() {
    uint32_t bIdx = 0;
    uint32_t width = REALTIME_JOINT_TEXTURE_WIDTH;
    uint32_t height = REALTIME_JOINT_TEXTURE_HEIGHT;
    for (const auto &texture : _realTimeJointTexture->textures) {
        auto *buffer = _realTimeJointTexture->buffer;
        auto *dst = buffer;
        auto *src = _dataArray[bIdx];
        uint32_t count = width;
        for (uint32_t i = 0; i < count; i++) {
            dst = buffer + (4 * i);
            memcpy(dst, src, 16);
            src = src + 4;
            dst = buffer + (4 * (i + width));
            memcpy(dst, src, 16);
            src = src + 4;
            dst = buffer + 4 * (i + 2 * width);
            memcpy(dst, src, 16);
            src = src + 4;
        }
        uint32_t buffOffset = 0;
        gfx::TextureSubresLayers layer;
        gfx::Offset texOffset;
        gfx::Extent extent{width, height, 1};
        gfx::BufferTextureCopy region{
            buffOffset,
            width,
            height,
            texOffset,
            extent,
            layer};
        auto *device = gfx::Device::getInstance();

        device->copyBuffersToTexture(reinterpret_cast<const uint8_t *const *>(&buffer), texture, &region, 1);
        bIdx++;
    }
}

void SkinningModel::releaseData() {
    if (!_dataArray.empty()) {
        for (auto *data : _dataArray) {
            CC_SAFE_DELETE_ARRAY(data);
        }
        _dataArray.clear();
    }
    CC_SAFE_DELETE(_realTimeJointTexture);
    if (!_buffers.empty()) {
        for (gfx::Buffer *buffer : _buffers) {
            CC_SAFE_DESTROY(buffer);
        }
        _buffers.clear();
    }
}

} // namespace cc
